sw/airborne/modules/sensors/airspeed_sdp3x.c

```diff
@@ -0,0 +1,223 @@
+
+  // calculate 8-bit checksum with polynomial 0x31 (x^8 + x^5 + x^4 + 1)
+  for (unsigned i = 0; i < size; i++) {
+    crc_value ^= (data[i]);
+
+    for (int bit = 8; bit > 0; --bit) {
+      if (crc_value & 0x80) {
+        crc_value = (crc_value << 1) ^ 0x31;
+
+      } else {
+        crc_value = (crc_value << 1);
+      }
+    }
+  }
+
+  // verify checksum
+  return (crc_value == checksum);
+}
+
+static void sdp3x_downlink(struct transport_tx *trans, struct link_device *dev)
+{
+  int16_t temp = (int16_t)(sdp3x.temperature * 10.f);
+  pprz_msg_send_AIRSPEED_MS45XX(trans,dev,AC_ID,
+                                &sdp3x.pressure,
+                                &temp,
+                                &sdp3x.airspeed);
+}
+
+void sdp3x_init(void)
+{
+  sdp3x.pressure = 0.;
+  sdp3x.temperature = 0;
+  sdp3x.airspeed = 0.;
+  sdp3x.pressure_scale = SDP3X_PRESSURE_SCALE;
+  sdp3x.pressure_offset = SDP3X_PRESSURE_OFFSET;
+  sdp3x.airspeed_scale = SDP3X_AIRSPEED_SCALE;
+  sdp3x.sync_send = SDP3X_SYNC_SEND;
+  sdp3x.initialized = false;
+
+  sdp3x_trans.status = I2CTransDone;
+  // setup low pass filter with time constant and 100Hz sampling freq
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_AIRSPEED_MS45XX, sdp3x_downlink); // FIXME
+#endif
+}
+
+void sdp3x_periodic(void)
+{
+  if (sdp3x_trans.status != I2CTransDone) {
+    return; // not ready
+  }
+
+  if (sdp3x.initialized) {
+    // Initiate next read
+    i2c_receive(&SDP3X_I2C_DEV, &sdp3x_trans, SDP3X_I2C_ADDR, 6);
+  }
+  else {
+    // Init sensor in continuous mode
+    sdp3x_trans.buf[0] = SDP3X_MODE >> 8;
+    sdp3x_trans.buf[1] = SDP3X_MODE & 0xFF;
+    i2c_transmit(&SDP3X_I2C_DEV, &sdp3x_trans, SDP3X_I2C_ADDR, 2);
+    sdp3x.initialized = true;
+  }
+}
+
+void sdp3x_event(void)
+{
+  /* Check if transaction is succesfull */
+  if (sdp3x_trans.status == I2CTransSuccess) {
+
+    if (sdp3x.initialized) {
+      // make local copy of buffer
+      uint8_t buf[6];
+      for (uint8_t i = 0; i < 6; i++) {
+        buf[i] = sdp3x_trans.buf[i];
+      }
+
+      // Check the CRC
+      if (!sdp3x_crc(&buf[0], 2, buf[2]) || !sdp3x_crc(&buf[3], 2, buf[5])) {
+        // error
+        sdp3x_trans.status = I2CTransDone;
+        return;
+      }
+
+      int16_t p_raw = ((int16_t)(buf[0]) << 8) | (int16_t)(buf[1]);
+      sdp3x.pressure = ((float)p_raw / sdp3x.pressure_scale) - sdp3x.pressure_offset;
+
+      int16_t t_raw = ((int16_t)(buf[3]) << 8) | (int16_t)(buf[4]);
+      sdp3x.temperature = (float)t_raw / SDP3X_SCALE_TEMPERATURE;
+
+      // Send (differential) pressure via ABI
+      AbiSendMsgBARO_DIFF(SDP3X_SENDER_ID, sdp3x.pressure);
+      // Send temperature as float in deg Celcius via ABI
+      AbiSendMsgTEMPERATURE(SDP3X_SENDER_ID, sdp3x.temperature);
+      // Compute airspeed
+      sdp3x.airspeed = sqrtf(Max(sdp3x.pressure * sdp3x.airspeed_scale, 0));
+
+#if USE_AIRSPEED_SDP3X
+      stateSetAirspeed_f(sdp3x.airspeed);
```

Comment:
Isn't it nicer to remove this and instead use the `air_data` module?

Reply:
I agree. I guess other sensors should be updated the same way, but I'm afraid that it will break some airframe files since air_data is not always loaded in "historical" config file.